Replace with_schema_unchecked with new_unchecked

tustvold opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #.

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

These schema checks exist for a reason, it should not be possible to bypass them without using unsafe, as allows constructing invalid arrays, via conversion to StructArray, which in turn could easily lead to UB.

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?

No this method was added yesterday and has not been released - https://github.com/apache/arrow-rs/pull/7402
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please call them out.
-->

alamb commented on arrow-array/src/record_batch.rs:
```diff
@@ -359,11 +359,12 @@ impl RecordBatch {
+    /// # Safety
+    ///
+    /// `schema` must be a superset of the current schema as determined by [`Schema::contains`]
+    pub unsafe fn with_schema_unchecked(self, schema: SchemaRef) -> Result<Self, ArrowError> {
```
I agree this API seems awkward and not consistent with the other APIs in this crate

What I suggest is:

1. Remove this API
2. Add APIs for destructing a RecordBatch into its constituent parts: 
3. Add an `unsafe` construction API (inline with the other parts of arrow)

So something like
```rust
impl RecordBatch {
  /// Creates a RecordBatch without any consistency checks
  pub unsafe fn new_unchecked(schema: Schema, arrays: vec<ArrayRef>, row_count) -> Self {
   ...
  }
  
  /// Return the inner fields of this RecordBatch, consuming sekf
  pub fn into_parts(self) -> (Schema, Vec<ArrayRef>, usize) {
   ...
  }
}
```

tustvold replied:
I read this as a performance optimisation, not wanting to remove the constraints full stop as #6855, which was rejected for being unsound

alamb replied:
BTW I also think we should add the following API. Can be a follow on

```rust
  /// Return the inner fields of this RecordBatch, consuming sekf
  pub fn into_parts(self) -> (Schema, Vec<ArrayRef>, usize) {
   ...
  }
```

comphead reviewed (approved):
Thanks @tustvold and @alamb for  making it unsafe is awesome idea to double emphasize risks. Similar to https://doc.rust-lang.org/std/primitive.slice.html#method.get_unchecked where bounds checks are not done by STD lib and moved to user caller site instead

tustvold commented:
As discussed in https://github.com/apache/arrow-rs/pull/7405#discussion_r2039312086 I have reverted with_schema_unchecked and added a new_unchecked constructor

alamb commented on arrow-array/src/record_batch.rs:
```diff
@@ -240,6 +241,27 @@ impl RecordBatch {
         Self::try_new_impl(schema, columns, &options)
     }
 
+    /// Creates a `RecordBatch` from a schema and columns
+    ///
+    /// # Safety
+    ///
+    /// Expects the following:
+    ///
+    ///  * `schema.fields.len() == columns.len()`
+    ///  * `schema.fields[i].data_type() == columns[i].data_type()`
+    ///  * `columns[i].len() == row_count`
```
```suggestion
    /// Creates a `RecordBatch` from a schema and columns, without validation.
    ///
    /// See [`Self::try_new`] for the checked version. 
    ///
    /// # Safety
    ///
    /// Expects the following:
    ///
    ///  * `schema.fields.len() == columns.len()`
    ///  * `schema.fields[i].data_type() == columns[i].data_type()`
    ///  * `columns[i].len() == row_count`
    ///
    /// Note: if the schema, does not match the underlying data exactly, this can lead to undefined behavior
    /// for example,  via conversion to a `StructArray`, which in turn could lead to incorrect access. 
```

alamb reviewed (approved):
Looks good to me

tustvold commented:
FYI I also considered using RecordBatchOptions instead of just row_count, however, the only other option is irrelevant to an unchecked constructor - and is also something we probably should remove - https://github.com/apache/arrow-rs/pull/7406
